ic-canister/ic-canister/src/lib.rs

```diff
@@ -119,3 +388,84 @@ pub trait Canister {
     /// Returns the principal of the canister.
     fn principal(&self) -> Principal;
 }
+
+type ResponderFn = dyn Fn(Vec<u8>) -> CallResult<Vec<u8>>;
+type ResponderHashMap = HashMap<(Principal, String), Box<ResponderFn>>;
+
+thread_local! {
+    static __RESPONDERS: Rc<RefCell<ResponderHashMap>> = Rc::new(RefCell::new(HashMap::new()));
+}
+
+fn _register_virtual_responder(
+    principal: Principal,
+    method_name: &str,
+    responder: impl Fn(Vec<u8>) -> CallResult<Vec<u8>> + 'static,
+) {
+    __RESPONDERS.with(|responders| {
+        responders
+            .borrow_mut()
+            .insert((principal, method_name.to_string()), Box::new(responder));
+    })
+}
+
+/// Invokes a virtual canister method. This function is supposed to be called through [virtual_canister_call] macro.
+pub fn call_virtual_responder(
```

Comment:
add a `#[doc(hidden)]` here since its not meant to be called directly by the user or prefix the name with an underscore or two?